src/CompiledMatcher.php

```diff
@@ -0,0 +1,71 @@
+<?php
+
+/*
+ * This file is part of composer/semver.
+ *
+ * (c) Composer <https://github.com/composer>
+ *
+ * For the full copyright and license information, please view
+ * the LICENSE file that was distributed with this source code.
+ */
+
+namespace Composer\Semver;
+
+use Composer\Semver\Constraint\Constraint;
+use Composer\Semver\Constraint\ConstraintInterface;
+use Composer\Semver\Constraint\NotCompilableConstraintException;
+
+/**
+ * Helper class to evaluate constraint by compiling and reusing the code to evaluate
+ */
+class CompiledMatcher
+{
+    private static $checked = array();
+    private static $enabled = null;
+
+    private static $transOpInt = array(
+        Constraint::OP_EQ => '==',
+        Constraint::OP_LT => '<',
+        Constraint::OP_LE => '<=',
+        Constraint::OP_GT => '>',
+        Constraint::OP_GE => '>=',
+        Constraint::OP_NE => '!=',
+    );
+
+    /**
+     * Evaluates the expression: $constraint match $operator $version
+     *
+     * @param ConstraintInterface $constraint
+     * @param int                 $operator
+     * @param string              $version
+     *
+     * @return mixed
+     */
+    public static function match(ConstraintInterface $constraint, $operator, $version)
+    {
+        if (self::$enabled === null) {
+            self::$enabled = !in_array('eval', explode(',', ini_get('disable_functions')));
+        }
+        if (!self::$enabled) {
+            return $constraint->matches(new Constraint(self::$transOpInt[$operator], $version));
+        }
+
+        $cacheKey = $operator.$constraint;
+        if (!isset(static::$checked[$cacheKey])) {
```

Comment:
in php < 8.0 using `\array_key_exists()` is faster then `isset()`

Reply:
are you sure about your sentence ? I always heard the opposite. So either you swapped the alternatives, or you swapped the version check.

Reply:
you are right.. mixed it up.

in php <8.0 the usual workaround is
```
if (isset($array[$key]) || \array_key_exists($key, $array)) {
```

while in php8+
```
if (\array_key_exists($key, $array)) {
```
is faster

Reply:
From every bench I see, that's the opposite for php <7.2. And since php>7.2 time are ~ equivalent
Note: the old trick replacing `array_key_exists(...)` by `isset(...) || array_key_exists(...)`  are not need anymore since 7.2 as both method have the same performance